opennav_following/src/following_server.cpp

```diff
@@ -0,0 +1,604 @@
+FollowingServer::on_shutdown(const rclcpp_lifecycle::State &)
+{
+  RCLCPP_INFO(get_logger(), "Shutting down %s", get_name());
+  return nav2::CallbackReturn::SUCCESS;
+}
+
+template<typename ActionT>
+void FollowingServer::getPreemptedGoalIfRequested(
+  typename std::shared_ptr<const typename ActionT::Goal> goal,
+  const typename nav2::SimpleActionServer<ActionT>::SharedPtr & action_server)
+{
+  if (action_server->is_preempt_requested()) {
+    goal = action_server->accept_pending_goal();
+  }
+}
+
+template<typename ActionT>
+bool FollowingServer::checkAndWarnIfCancelled(
+  typename nav2::SimpleActionServer<ActionT>::SharedPtr & action_server,
+  const std::string & name)
+{
+  if (action_server->is_cancel_requested()) {
+    RCLCPP_WARN(get_logger(), "Goal was cancelled. Cancelling %s action", name.c_str());
+    return true;
+  }
+  return false;
+}
+
+template<typename ActionT>
+bool FollowingServer::checkAndWarnIfPreempted(
+  typename nav2::SimpleActionServer<ActionT>::SharedPtr & action_server,
+  const std::string & name)
+{
+  if (action_server->is_preempt_requested()) {
+    RCLCPP_WARN(get_logger(), "Goal was preempted. Cancelling %s action", name.c_str());
+    return true;
+  }
+  return false;
+}
+
+void FollowingServer::followObject()
+{
+  std::lock_guard<std::mutex> lock(*dynamic_params_lock_);
+  action_start_time_ = this->now();
+  rclcpp::Rate loop_rate(controller_frequency_);
+
+  auto goal = following_action_server_->get_current_goal();
+  auto result = std::make_shared<FollowObject::Result>();
+
+  if (!following_action_server_ || !following_action_server_->is_server_active()) {
+    RCLCPP_DEBUG(get_logger(), "Action server unavailable or inactive. Stopping.");
+    return;
+  }
+
+  if (checkAndWarnIfCancelled<FollowObject>(following_action_server_, "follow_object")) {
+    following_action_server_->terminate_all();
+    return;
+  }
+
+  getPreemptedGoalIfRequested<FollowObject>(goal, following_action_server_);
+  num_retries_ = 0;
+
+  try {
+    auto target_frame = goal->frame_id;
+    if (!target_frame.empty()) {
+      RCLCPP_INFO(get_logger(), "Following frame: %s instead of pose", target_frame.c_str());
+    }
+
+    // Following control loop: while not timeout, run controller
+    geometry_msgs::msg::PoseStamped object_pose;
+    auto start = this->now();
+    rclcpp::Duration max_duration = goal->max_duration;
+    while (rclcpp::ok()) {
+      try {
+        // Check if we have run out of time
+        if (this->now() - start > max_duration && max_duration.seconds() > 0.0) {
+          RCLCPP_INFO(get_logger(), "Exceeded max duration. Stopping.");
+          result->total_elapsed_time = this->now() - action_start_time_;
+          result->num_retries = num_retries_;
+          publishZeroVelocity();
+          following_action_server_->succeeded_current(result);
+          return;
+        }
+
+        // Approach the object using control law
+        if (approachObject(object_pose, target_frame)) {
```

Comment:
I think another timeout here would be good. If the obstacle has been still for > N seconds, we should succeed the action and return

Reply:
Mmm, wouldn't this timeout override the max_duration timeout? The ```max_duration``` timeout is the maximum time during which an object will be tracked. If another timeout is added here then, when the object is reached, after this timeout, it will return 'succeed', but there will still be time available as ```new_timeout < max_duration```.